# Analyse de la qualité des données

## Prise de connaissance avec le jeu de données

### Chargement des données brut

In [1]:
from getting_started import df_patient, df_pcr

Conversion de chaque attribut du référentiel en un type de données plus spécifique.

In [2]:
df_patient = df_patient.convert_dtypes()

df_patient.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   patient_id     20000 non-null  Int64 
 1   given_name     19560 non-null  string
 2   surname        19575 non-null  string
 3   street_number  19618 non-null  Int64 
 4   address_1      19204 non-null  string
 5   suburb         19788 non-null  string
 6   postcode       19801 non-null  string
 7   state          18010 non-null  string
 8   date_of_birth  17989 non-null  Int64 
 9   age            16003 non-null  Int64 
 10  phone_number   19081 non-null  string
 11  address_2      7893 non-null   string
dtypes: Int64(4), string(8)
memory usage: 1.9 MB


Conversion de chaque attribut de l'échantillon en un type de données plus spécifique.

In [3]:
df_pcr = df_pcr.convert_dtypes()

df_pcr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8800 entries, 0 to 8799
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   patient_id  8800 non-null   Int64 
 1   pcr         8800 non-null   string
dtypes: Int64(1), string(1)
memory usage: 146.2 KB


### Présence de doublons sur l'identifiant

In [4]:
with_duplicated_id = df_patient.patient_id.duplicated(keep=False)

df_patient[with_duplicated_id].sort_values("patient_id")

patient_id given_name        surname  street_number  \
12117      109304    zachary      farronato             30   
14839      109304     bailey      donaldson             20   
4386       110207       toby          brock              4   
12989      110207       zali          brock             32   
10184      115791     hannah         clarke             70   
...           ...        ...            ...            ...   
10507      990695       erin       braunack             49   
8764       990936        amy          royle             90   
12563      990936   samantha          green             21   
2385       994235      trent  stewart-jones            129   
7961       994235     teresa      chowdhury             60   

                 address_1        suburb postcode state  date_of_birth   age  \
12117       outtrim avenue          como     2196   vic       19090801    31   
14839       tardent street          ryde     0812   qld       19580310    26   
4386     merriman crescent      baralaba     3025   nsw       19000424    35   
12989        hedger street        toorak     5038   act           <NA>    22   
10184      galmarra street      mayfield     7010   vic       19830828    25   
...                    ...           ...      ...   ...            ...   ...   
10507     moondarra street   broken hill     2640   qld       19830122    30   
8764     whittell crescent       coramba     5032    sa       19950326  <NA>   
12563       brierly street     ardrossan     2140  <NA>       19380210    29   
2385   macfarland crescent    wangaratta     2732   nsw           <NA>  <NA>   
7961                  <NA>  chester hill     2147   vic       19010407    27   

      phone_number       address_2  
12117  07 22894061       the reefs  
14839  07 13479210            <NA>  
4386   08 33842007         leitrim  
12989  08 96818512            <NA>  
10184  04 70760611            <NA>  
...            ...             ...  
10507  03 69523317         yuulong  
8764   08 07309295  tewantin plaza  
12563  02 51600621            <NA>  
2385   07 98662458       mountview  
7961   02 83271460            <NA>  

[403 rows x 12 columns]

Il y a 403 patients qui partagent un identifiant.

### Présence de doublons sur l'ensemble des attributs

In [5]:
with_same_attributes = df_patient.drop(columns="patient_id").duplicated(keep=False)

df_patient[with_same_attributes].sort_values(by=["surname", "given_name"])

patient_id given_name   surname  street_number         address_1  \
6981       804259    charlie    gamlin             33    nungara street   
12438      107928    charlie    gamlin             33    nungara street   
15184      744576      freya   jaffres             50        hart place   
17522      449738      freya   jaffres             50        hart place   
5729       973030    delaney   kermeen              9      wallen place   
15337      373129    delaney   kermeen              9      wallen place   
1052       664037   samantha    laundy              2   mannheim street   
18867      421721   samantha    laundy              2   mannheim street   
1863       658924      lewis  matthews             24     allawah flats   
1944       669936      lewis  matthews             24     allawah flats   
942        943523     taliah    monaco             15  national circuit   
1769       437054     taliah    monaco             15  national circuit   
8900       395901       <NA>     polak             35        jess place   
16552      279300       <NA>     polak             35        jess place   
4440       220263      ebony  rawlings             83      spica street   
9421       950181      ebony  rawlings             83      spica street   
4819       396208       kyle  sherriff              5   jim pike avenue   
9724       794186       kyle  sherriff              5   jim pike avenue   
5393       887301        ava      tsun             75  twamley crescent   
17752      122873        ava      tsun             75  twamley crescent   
8228       834172  gabrielle      <NA>             56    langdon avenue   
17344      369594  gabrielle      <NA>             56    langdon avenue   

               suburb postcode state  date_of_birth   age phone_number  \
6981        bayswater     5251   qld       19190111    24  04 33326042   
12438       bayswater     5251   qld       19190111    24  04 33326042   
15184  biggera waters     4413    wa           <NA>    33  07 76055136   
17522  biggera waters     4413    wa           <NA>    33  07 76055136   
5729    st kilda east     4405   qld       19391226    25  08 37919311   
15337   st kilda east     4405   qld       19391226    25  08 37919311   
1052       quairading     4740   qld       19480111  <NA>  02 37735421   
18867      quairading     4740   qld       19480111  <NA>  02 37735421   
1863          attunga     3216    wa       19690623    26  03 95122427   
1944          attunga     3216    wa       19690623    26  03 95122427   
942          hillside     4211   nsw       19171005    37         <NA>   
1769         hillside     4211   nsw       19171005    37         <NA>   
8900        annandale     2250   qld       19531130  <NA>  03 33180608   
16552       annandale     2250   qld       19531130  <NA>  03 33180608   
4440     albany creek     3156   qld           <NA>    30         <NA>   
9421     albany creek     3156   qld           <NA>    30         <NA>   
4819     bossley park     6111   nsw       19480823  <NA>  02 22087371   
9724     bossley park     6111   nsw       19480823  <NA>  02 22087371   
5393            eaton     5109   nsw       19231129    36  07 70215543   
17752           eaton     5109   nsw       19231129    36  07 70215543   
8228      castle cove     2564   qld       19350604    14  02 92453197   
17344     castle cove     2564   qld       19350604    14  02 92453197   

             address_2  
6981              <NA>  
12438             <NA>  
15184             <NA>  
17522             <NA>  
5729              <NA>  
15337             <NA>  
1052           stanton  
18867          stanton  
1863              <NA>  
1944              <NA>  
942               <NA>  
1769              <NA>  
8900       bonnie doon  
16552      bonnie doon  
4440       sunny braes  
9421       sunny braes  
4819   cuthberts court  
9724   cuthberts court  
5393              <NA>  
17752             <NA>  
8228              <NA>  
17344             <NA>

Il y 22 patients qui sont des doublons parfaits hors identifiant.

Ces deux dernières observations justifient une analyse approfondie sur le référenciel de patients afin de définir une stratégie de dédoublonnage et de réconciliation de données adaptée en préambule de l'analyse exploratoire.

## Analyse de l'échantillon de tests PCR

### Répartition des valeurs du test PCR

In [6]:
df_pcr.pcr.value_counts()

N           3482
Negative    3134
Positive    1283
P            901
Name: pcr, dtype: Int64

Deux conventions sont utilisées pour représenter les deux valeurs possibles d'un test PCR (négatif ou positif) : `N / P` et `Negative / Positive`.

Il faudra normaliser ces résultats dans une variable catégorielle ordonnée.

### Exhaustivité du référentiel

In [7]:
df_pcr.patient_id.isin(df_patient.patient_id).all()

True

L'intégralité des identifiants associés à chaque test de l'échantillon sont présents dans le référenciel.

## Analyse du référenciel de patients

### Nom et prénom

#### Valeurs manquantes

In [8]:
df_na_in_patient_name = df_patient[["surname", "given_name"]].isna()

df_na_in_patient_name.value_counts()

surname  given_name
False    False         19139
         True            436
True     False           421
         True              4
dtype: int64

Il y a 861 patients dont le nom et / ou prénom ne sont pas renseignés.

Répartition des valeurs :

In [9]:
df_patient_full_name = df_patient[["surname", "given_name"]].dropna()

df_patient_full_name.value_counts()

surname   given_name
white     emiily        14
green     emiily        12
white     joshua        11
campbell  joshua        11
ryan      emiily        10
                        ..
newberry  jack           1
          hannah         1
          dominic        1
          daniel         1
aaberg    charlotte      1
Length: 16681, dtype: int64

#### Fautes typographiques

In [22]:
from jellyfish import damerau_levenshtein_distance
from pandas import merge

df_patient["full_name"] = df_patient.agg(
    lambda x: f"{x.given_name} {x.surname}", axis="columns")

df_full_name = df_patient[["patient_id", "full_name", "phone_number"]].dropna()
df_full_name = df_full_name.merge(df_full_name, on="phone_number")
df_full_name = df_full_name[df_full_name.patient_id_x != df_full_name.patient_id_y]
df_full_name["linked_ids"] = df_full_name[["patient_id_x", "patient_id_y"]].apply(
    lambda x: tuple(sorted(x)), axis="columns")
df_full_name.drop_duplicates("linked_ids", inplace=True)
df_full_name["similarity"] = df_full_name.apply(
    lambda row: damerau_levenshtein_distance(row.full_name_x, row.full_name_y), axis="columns")

df_full_name[df_full_name.similarity == 1].sort_values("phone_number").head(10)

patient_id_x     full_name_x phone_number  patient_id_y     full_name_y  \
4192        311830     taaila <NA>  02 00325977        210155     taalia <NA>   
4193        311830     taaila <NA>  02 00325977        525466     taalia <NA>   
985         431593   adam schumann  02 01272164        123387  adam schumajnn   
991         123387  adam schumajnn  02 01272164        375877   adam schumann   
990         123387  adam schumajnn  02 01272164        505218   adam schumann   
1827        489678   jacob svenson  02 03546747        909797    jaob svenson   
1831        669354   jacob svenson  02 03546747        909797    jaob svenson   
1835        576055   jacob svenson  02 03546747        909797    jaob svenson   
579         399260  <NA> peterssen  02 03687263        776651   <NA> petersen   
574         950122   <NA> petersen  02 03687263        399260  <NA> peterssen   

            linked_ids  similarity  
4192  (210155, 311830)           1  
4193  (311830, 525466)           1  
985   (123387, 431593)           1  
991   (123387, 375877)           1  
990   (123387, 505218)           1  
1827  (489678, 909797)           1  
1831  (669354, 909797)           1  
1835  (576055, 909797)           1  
579   (399260, 776651)           1  
574   (399260, 950122)           1

In [23]:
df_full_name[df_full_name.similarity > 1].sort_values("phone_number").head(10)

patient_id_x        full_name_x phone_number  patient_id_y  \
6144         953966         to godfrey  02 01871708        405442   
16219        382081         james <NA>  02 03755662        687453   
589          471602    briekle tippuns  02 04356284        106764   
585          718467    brielle tippins  02 04356284        471602   
5705         496474      stacia seddon  02 05657798        963018   
5702         970678      stacia seddon  02 05657798        963018   
4075         131348    timothy britten  02 08766786        646113   
10877        898204  jordan ballantyne  02 13710140        494861   
10881        740337  jordan ballantyne  02 13710140        494861   
10882        740337  jordan ballantyne  02 13710140        219095   

             full_name_y        linked_ids  similarity  
6144      thomas godfrey  (405442, 953966)           4  
16219           jim <NA>  (382081, 687453)           3  
589      brielle tippins  (106764, 471602)           2  
585      briekle tippuns  (471602, 718467)           2  
5705    anastasia seddon  (496474, 963018)           4  
5702    anastasia seddon  (963018, 970678)           4  
4075      timothy bripen  (131348, 646113)           2  
10877  ballantyne jordan  (494861, 898204)          14  
10881  ballantyne jordan  (494861, 740337)          14  
10882    <NA> ballantyne  (219095, 740337)           6

Valeurs manquantes :

In [ ]:
df_patient[["age", "date_of_birth"]].isna().value_counts()

Dates de naissance invalides :

In [ ]:
from pandas import to_datetime

df_patient["dob_datetime"] = to_datetime(
    df_patient.date_of_birth,
    errors="coerce",
    format="%Y%M%d",
)

df_patient[["age", "date_of_birth", "dob_datetime"]].sample(10, random_state=42)

Incohérences entre âge déclaré et calculé : 

In [ ]:
from datetime import datetime
df_patient["age_from_dob"] = ((datetime.today() - df_patient["dob_datetime"].dropna()).dt.days).floordiv(365)

df_patient.head(20)

## Analyse des états et codes postaux

Valeurs manquantes

In [ ]:
df_patient[["state", "postcode"]].isna().value_counts()

Valeurs des états

In [ ]:
states = df_patient.state.dropna().value_counts()

print(states[:8])
print(states[8:])
print(states[8:].sum())

Incohérence état / postcode

In [ ]:
postcodes = df_patient.postcode.dropna().value_counts()

Inversion postcode / quartier

In [ ]:
df_patient.loc[df_patient.suburb.str.isnumeric()]

## Sanitize postcode and state

Assume postcode is more reliable than state.
Test all postcodes are valid.
Case postcode invalid, try swap with suburb.
Test some state are invalid.
Normalize state with typos.
For missing or invalid states, guess from postcode.
Keep state if postcode invalid.


In [ ]:
df_patient.state = df_patient.state.str.upper()
df_patient_invalid_postcode = df_patient[df_patient.state.isin(["NSW", "VIC", "QLD", "WA", "SA", "TAS", "ACT", "NT"])]
df_patient_vic = df_patient_invalid_postcode.loc[df_patient_invalid_postcode.state == "VIC"]
df_patient_vic.loc[~df_patient_vic.postcode.str.match(r"[3|8]\d{3}")].head()

## Sanitize phone numbers

- Format XX XXXXXXXX, 04 if mobile
- Must be coherent with postcode


In [ ]:
df_patient.loc[~df_patient.phone_number.str.match(r"\d{2}\s\d{8}")].phone_number.count()